enhancements/results-forwarding.md

```diff
@@ -0,0 +1,513 @@
+    tls:
+      dynamic: 
+        caSecretName: $caSecretName
+        serverSecretName: $ref
+        clientSecretName: $ref
+    authentication:
+      type: mtls
+    extraMetadata:
+      clusterName: foo
+      clusterType: bar
+roles:
+- worker 
+- master 
+scanTolerations: 
+  default:
+  - operator: Exists
+  schedule: 0 1 * * * 
+```
+
+The following forwards results to a gRPC endpoint using token authentication.
+It also sets additional metadata for the gRPC implementation.
+
+```yaml
+apiVersion: compliance.openshift.io/v1alpha1
+kind: ScanSetting
+metadata:
+  name: default
+  namespace: openshift-compliance
+rawResultStorage:
+  pvAccessModes:
+  - ReadWriteOnce 
+  rotation: 3 
+  size: 1Gi 
+resultForwarding:
+  provider: grpc
+  grpc:
+    version: v1
+    endpoint: https://compliance-endpoint.example.com
+    tls:
+      dynamic: 
+        caSecretName: $caSecretName
+    authentication:
+      type: token
+      token:
+        tokenSecretName: $ref
+    extraMetadata:
+      clusterName: foo
+      clusterType: bar
+      randomKey: random value
+roles:
+- worker 
+- master 
+scanTolerations: 
+  default:
+  - operator: Exists
+  schedule: 0 1 * * * 
+```
+
+#### gRPC API
+
+This section focuses solely on the details for implementing a forwarder for
+gRPC.
+
+```protobuf
+message Result {
+  string subject = 1;
+  string control = 2;
+  string rule = 3;
+  string assessment_id = 4;
+  string outcome = 5;
+  string description = 6;
+  string severity = 7;
+  string instructions = 8;
+}
+```
+
+Example:
+
+* `subject`: `cluster.example.com`
+  - A unique name for the cluster, may come from the `ScanSetting` CRD
+* `control`: `NIST-800-53: CM-6;CM-6(1)`
+  - A list of controls this check is applicable to
+* `rule`: `ocp4-api-server-admission-control-plugin-alwayspullimages`
+  - The rule name as defined in the ComplianceAsCode/content
+* `assessment_id`: `108268c9-af42-4004-bf43-2729648d63da`
+  - The unique ID of the scan performed by the operator
+* `outcome`: `PASS`
+  - The overall status of the rule (e.g., `PASS`, `FAIL`, `MANUAL`)
+* `description`: `Ensure that the Admission Control Plugin AlwaysPullImages is not set ...`
+  - Description as defined in the ComplianceAsCode/content
+* `severity`: `high`
+  - How severe this finding is, if it's set by the ComplianceAsCode/content
+* `instructions`: `Run the following command: $ oc -n openshift-kube-apiserver ...`
+  - Instructions as defined in the ComplianceAsCode/content
+
+### Implementation Details/Notes/Constraints
+
+A first step in implementing the forwarding proposal would be to work with the existing
+`v1alpha1` `ScanSetting` object. All changes would need to be done in how the Compliance
+Operator effectuates scans, and thus we'd ensure that backwards compatibility is
+taken into account. A new `crd-creator` pod would be introduced to receive the results
+from the `aggregator` and ensure that the known `ComplianceCheckResult` and
```

Comment:
This approach makes it so that we'll need to network client-ize the aggregator, and the crd-creator is a server.  There are other alternatives that we might prefer
1. Keep aggregator as it is, and when forwarding is configured, an additional workload is spawned for forwarding. The `crd-forwarder` waits for the aggregator to make results and remediation CRs, and then forwards. (the downside is that local results are always needed)  
2. Add forwarder and keepalive routines to aggregator that run when forwarding is configured.

Reply:
> This approach makes it so that we'll need to network client-ize the aggregator, and the crd-creator is a server. There are other alternatives that we might prefer
> 
>     1. Keep aggregator as it is, and when forwarding is configured, an additional workload is spawned for forwarding. The `crd-forwarder` waits for the aggregator to make results and remediation CRs, and then forwards. (the downside is that local results are always needed)
> 

What's, in your opinion, preferable about scraping the CRs in your opinion as opposed to the client/server model? Are you concerned about the kind of issues like we had with certificates for resultserver?

>     2. Add forwarder and keepalive routines to aggregator that run when forwarding is configured.

Sorry, what is keepalive routines?

I thought that one of the reasons to have the forwarder as client-server was that each consumer of the forwarder (one being our compliance service) would implement the server part, then CO doesn't have to care about how the compliance service consumes the forwarded results, it just need to be able to send the results.

Reply:
> 



> > This approach makes it so that we'll need to network client-ize the aggregator, and the crd-creator is a server. There are other alternatives that we might prefer
> > ```
> > 1. Keep aggregator as it is, and when forwarding is configured, an additional workload is spawned for forwarding. The `crd-forwarder` waits for the aggregator to make results and remediation CRs, and then forwards. (the downside is that local results are always needed)
> > ```
> 
> What's, in your opinion, preferable about scraping the CRs in your opinion as opposed to the client/server model? Are you concerned about the kind of issues like we had with certificates for resultserver?
> 

That's mainly it, by adding a new network link between the scan pods and the CRDs, we introduce some possible unreliability (at least if we use certificates). If we did go down this path we should at least think about using the token option, and we wouldn't have to worry about certs timing out if scans or transmissions are slow.

> > ```
> > 2. Add forwarder and keepalive routines to aggregator that run when forwarding is configured.
> > ```
> 
> Sorry, what is keepalive routines?

I was thinking of keepalive routines as in, async making sure that the configured endpoint is available (part of the client's code). 

> 
> I thought that one of the reasons to have the forwarder as client-server was that each consumer of the forwarder (one being our compliance service) would implement the server part, then CO doesn't have to care about how the compliance service consumes the forwarded results, it just need to be able to send the results.

Yeah, for sure we'd need the client implementation (whether that is part of the aggregator or another workload). It's having a default server implementation that we use even when not forwarding (the mentioned `crd-creator`) that I would question.

Reply:
I can see your point. I still think that having the aggregator as a dumb client forwarding to `crd-creator1 is architecturally cleaner (no local-only path) but you do have a point that the local-only patch would probably provide fewer headaches (read: customer cases)